### PR TITLE
Add social login support for JWT authenfication

### PR DESCRIPTION
**Pull request by moifort**
* Add new question to enable social login

How I implement, step by step:
1. Authenticate to social login external server (i.e. Google) 
2. Authenticate to the JHipster server
3. JHipster server send request to angular with a cookie containing the jwt token
4. Angular controller store the token and valid the session

Linked to #2848

### CODE REVIEW

**Comment by deepu105** [comment]
Does twitter still works?

**Comment by moifort** [comment]
I didn't check twitter yet.

**Comment by moifort** [comment]
:boom: all the test are failed. Go to sleep...

**Comment by mraible** [comment]
Just curious - why a cookie instead of a header or body? The `UserJWTController` uses both header and body (for Angular 2).

https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/main/java/package/web/rest/_UserJWTController.java#L43

**Comment by moifort** [comment]
I try to make like the normal way, but I think social sign in cannot use the normal way to connect the app and need to have a special login authentication. There are my problems:

1. I try to use `api/authenticate` to authenticate the user, but the problem you need a non encrypted password to authorize the user. In my case, my password is already encoded, because user was already created by social sign in.

2. The social sign in mechanism is redirects urls (I didn't success to manage it with $http), and it's finish (if the social connection was successful) by sending a sign in response to the app, customizable through `CustomSignInAdapter.java`. I can custom my response request with `HttpServletResponse` object, this one can set header but not body. I try to add it in the header, but the problem, is how to get the header with a angular controller through ui-router, I didn't find a way. I have an other solution, catch it with an interceptor, but for me, it's a little to heavy.

My solution was to send a token in a expirable cookie and allow angular controller to use a login with token to authenticate the app. 

I don't think is the best solution, but I didn't find other. Maybe I miss some point or mechanism. So if you have other solution or implementation, tell me!

File: generators/server/templates/src/main/java/package/security/social/_CustomSignInAdapter.java [hunk 1]
```diff
@@ -10,17 +14,57 @@
 import org.springframework.social.connect.Connection;
 import org.springframework.social.connect.web.SignInAdapter;
 import org.springframework.web.context.request.NativeWebRequest;
+<%_ if (authenticationType == 'jwt') { _%>
+import com.mycompany.myapp.security.jwt.TokenProvider;
```
Comment (amitjindal) [inline generators/server/templates/src/main/java/package/security/social/_CustomSignInAdapter.java]: This import is incorrect.

Comment (moifort) [inline generators/server/templates/src/main/java/package/security/social/_CustomSignInAdapter.java]: :+1: well seen!

**Comment by moifort** [comment]
@deepu105 I will need some help, I try to understand the error on Travis. I download the `.yo-rc` who test the social and I didn't reproduce the error, I think I miss something but I have no clue...

**Comment by deepu105** [comment]
Ok ill take a look at the error

File: generators/server/index.js [hunk 2]
```diff
@@ -250,6 +246,25 @@ module.exports = JhipsterServerGenerator.extend({
                 },
                 {
                     when: function (response) {
+                        return applicationType == 'monolith' && (response.authenticationType == 'session' || response.authenticationType == 'jwt');
+                    },
+                    type: 'list',
+                    name: 'enableSocialSignIn',
```
Comment (pascalgrimaud) [inline generators/server/index.js]: you have to change the current tests 
https://github.com/jhipster/generator-jhipster/blob/master/test/test-creation.js#L676-L692

Comment (moifort) [inline generators/server/index.js]: :+1: @pascalgrimaud, I will add test for JWT social too!

**Comment by moifort** [comment]
It's seems working (#2848), maybe someone want to make some remarks or improvement? otherwise we can merge it.

**Comment by deepu105** [comment]
Was the two issues reported by amit solved or do we need to track it in a
new issue?

**Comment by deepu105** [comment]
Also a small doubt does it work with non email usernames?

**Comment by deepu105** [comment]
@jdubois can we merge this?

**Comment by moifort** [comment]
Need to check some little things it will be done tonight. 

Thibaut.

**Comment by moifort** [comment]
@deepu105 I just analyze the problem with the user delete. Since we removed the `login` rule on the entity `User`, in user management screen, the delete action on user use the login and spring rest had some problem to parse a query param with an email address format. I dont know about that, but maybe it's better to use id instead of login to update/delete/get an user?

**Comment by amitjindal** [comment]
@moifort I also checked and found that when the DELETE url is:
`http://localhost:8080/api/users/amitjindal@domain.com?cacheBuster=1458063515416`

In the `UserResource::deleteUser(@PathVariable String login)`
here the login is being set to: `amitjindal@domain`

So the delete is not working.


Also in this function we need to delete from `SocialUserConnection`

**Comment by moifort** [comment]
@amitjindal I think I will make a PR to fix that, but before I need to know if we can use `id` instead of `login`.

**Comment by deepu105** [comment]
Since the user entity id would be unique anyway I dont see any problem in
using that for delete and update

**Comment by moifort** [comment]
Thanks @deepu105 I create an issue to discuss about it. We can merge this PR, I will make the fix in another PR.

File: generators/server/index.js [hunk 3]
```diff
@@ -283,6 +279,25 @@ module.exports = JhipsterServerGenerator.extend({
                 },
                 {
                     when: function (response) {
+                        return applicationType == 'monolith' && (response.authenticationType == 'session' || response.authenticationType == 'jwt');
```
Comment (deepu105) [inline generators/server/index.js]: @moifort why did you add this as a new question, the way it was done earlier seemed better to me as its a type of auth and should be part of the same question

Comment (deepu105) [inline generators/server/index.js]: Oh ignore me you are enbling it for JWT i forgot

**Comment by fmatuszewski** [comment]
Since microservice gateway has also jwt authentication type is it possible to merge it there?

**Comment by fmatuszewski** [comment]
I checked the source code. It is the same codebase so updating enableSocialSignIn whould do the trick I suppose. So the only thing that is needed to add option in wizard(form).

**Comment by chegola** [comment]
When generate Microservices gateway, there is no option to choose Spring Social but I have manually added enableSocialSignIn: true in yo-rc.json then re-generate. jhipster generate social related files and it works great. Thanks you guys.

**Comment by gnoopy** [comment]
@chegola  May I double check with you?  Recently I've been curious about the social login feature with oauth2 configuration. Since there's no option on the wizard I think social login works only with JWT configuration.

Based on your comment,  you probably use oauth2 to enable microservices. Under this configuration you could manage to use social signin feature and no issues to use it. And this is just possible by modifying yo-rc.json and re-generate even though the wizard doesn't show the social login option with oauth2.  Is this correct on your experience ?

**Comment by fmatuszewski** [comment]
That is correct.
As far as I remember you also need to make some changes in the code
yourserlf. But I tried it on Angular js version.
I didn't try this with Angular 4 version.


Filip Matuszewski
604 59 99 36


2017-08-28 10:05 GMT+02:00 Francis <notifications@github.com>:

**Comment by gnoopy** [comment]
@fmatuszewski  thanks for your kind information. I will try this.

**Comment by fmatuszewski** [comment]
Let me know if you succeeded.

Filip Matuszewski
604 59 99 36


2017-08-29 2:01 GMT+02:00 Francis <notifications@github.com>:
